File: CPP/Implementation2_thread_safe/ellipse.cpp
#include "ellipse.h"
ellipse::ellipse()
{}
ellipse::ellipse(const ellipse& t)
{
	p = t.p;
}

position ellipse::get_position() 
{
	return p;
}
ellipse* ellipse::set_position(const position& pos)
{
	p =pos;
	return this;
}

void ellipse::move_to(position where)
{
	ms.move_to(this,where);
}
void ellipse::move_to(int dx, int dy)
{

	ms.move_to(this,dx,dy);
}
void ellipse::move_by(int dx, int dy)
{
	ms.move_by(this,dx,dy);
}	
void ellipse::move_by(position where)
{	
	ms.move_by(this,where);
}

File: JAVA/Implementation2_thread_safe/Ellipse.java
// One of geometric classes
public class Ellipse implements Movable {
	// Position of the geometric object on some canvas
	private Position p;
	private MoveServant m;
	
	public Ellipse()
	{
		m = new MoveServant();
	}
	
	// Method, which sets position of geometric object
	public void setPosition(Position p) {
		this.p = p;
	}

	// Method, which returns position of geometric object
	public Position getPosition() {
		return this.p;
	}
	
	public void moveTo(int dx, int dy)
	{
		m.moveTo(this,dx,dy);
	}
	public void moveTo(Position p)
	{
		m.moveTo(this,p);
	}
	public void moveBy(Position where)
	{
		m.moveBy(this,where);
	}
	public void moveBy(int dx, int dy)
	{
		m.moveBy(this,dx,dy);
	}
}